dateutil/parser/isoparser.py

```diff
@@ -0,0 +1,384 @@
+# -*- coding: utf-8 -*-
+"""
+This module offers a parser for ISO-8601 strings
+
+It is intended to support all valid date, time and datetime formats per the
+ISO-8601 specification, with a stricter mode for the most common subset.
+"""
+from datetime import datetime, timedelta, time, date
+import calendar
+from dateutil import tz
+
+import re
+
+__all__ = ["isoparse", "Isoparser"]
+
+
+class Isoparser(object):
```

Comment:
Is this something you expect users to subclass regularly?

Reply:
This is kinda one of the open questions, honestly. I've been a bit haphazard about the design of this class because I keep vacillating between "users will create an instance" and "users will create a subclass", which is why half the methods are `classmethod` and half are instance methods.

I think I'm partially being biased by the parser, where my hope was that downstream users would subclass `parser` to allow customization of behavior, but even in the parser, I'm thinking that might not be the right design for this particular sort of task.

After giving it a bit of thought, I believe that since ISO-8601 has a very clear set of options it supports, it might be best if it we refactored this to be more like `parser`/`parserinfo`, where `isoparserinfo` would have some limited feature flags like `ordinal_dates` and `weeks` (so that there's not this artificial division between "common" and "uncommon").

That said, in terms of subclasses, I think that when #449 is done, I'll probably want an internal subclass that returns a `PartialDatetime` instead of `datetime` (that way the `default_year` logic will be unnecessary).

Reply:
> [...] but even in the parser, I'm thinking that might not be the right design for this particular sort of task.

Yah, I've gone along with the make-everything-a-method approach since there isn't a _ton_ of downside, but I think YAGNI probably applies unless we have good reason to believe users actually are subclassing.

(This may be a discussion that belongs elsewhere, don't want to derail this PR's discussion)

Especially for e.g. `parser._build_tzinfo`, I have a hard time imagining the use case where it is easier to subclass than it is to pass a callable for `tzinfos`.  So I might start pushing to be more selective about does-this-really-need-to-be-a-method.

Reply:
> In any case, that doesn't mean I want bare functions sitting around.

OK, I evidently mis-interpreted your previous comment.  Never mind.